### ISSUE TITLE
How to get client's real ip?

### ISSUE THREAD

**Issue by bangbaew** [gofiber/contrib#507]
![image](https://user-images.githubusercontent.com/29559559/224698354-f8562f12-8686-4ddf-a988-e95d3a358045.png)

When make a remote request to my Gofiber endpoint, it gives http.client_ip = 10.8.11.189, which is container's local ip, but in Rust version of opentelemetry, used with Actix Web, it gives my real public ip out there, how can I make Gofiber's otel show public client ip?

**Comment by gaby** [comment]
@bangbaew Is the rust version also running inside Docker?

**Comment by gaby** [comment]
We can probably solve this by using this: https://docs.gofiber.io/api/ctx#ips

**Comment by bangbaew** [comment]
> > > @bangbaew Is the rust version also running inside Docker?
> > 
> > 
> > It's running inside a container, same network as the Gofiber app. This is the Rust library i use: https://github.com/OutThereLabs/actix-web-opentelemetry
> 
> Yeah, this is a Fiber bug.

Yeah, the log IPs on the terminal as well, they all are local IPs, and I don't think they're any useful.
![image](https://user-images.githubusercontent.com/29559559/224710447-0e3f8955-7e95-496b-8170-adfe136c5b67.png)

**Comment by gaby** [comment]
> > > > @bangbaew Is the rust version also running inside Docker?
> > > 
> > > 
> > > It's running inside a container, same network as the Gofiber app. This is the Rust library i use: https://github.com/OutThereLabs/actix-web-opentelemetry
> > 
> > 
> > Yeah, this is a Fiber bug.
> 
> Yeah, the log IPs on the terminal as well, they all are local IPs, and I don't think they're any useful. ![image](https://user-images.githubusercontent.com/29559559/224710447-0e3f8955-7e95-496b-8170-adfe136c5b67.png)

Those are expected since thats your IP inside the container. They only way to get the real IP in the logs is by parsing the Forwarded headers, it should be the first one in the List.

In one of your routes log `ctx.IPs()`

**Comment by ReneWerner87** [comment]
https://github.com/gofiber/contrib/blob/bae3c8cc2db1c8cfcf747203af34f0ffa6a96bd9/otelfiber/semconv.go#L59-L62

https://github.com/gofiber/fiber/blob/634f163e3f6292e658e61d0dd9e3c475d87b5d54/ctx.go#L699-L701

https://docs.gofiber.io/next/api/fiber#config
![image](https://user-images.githubusercontent.com/7063188/224712866-2b6efe52-e7f7-461c-853a-73b234e3323c.png)

did you configure this header ? otherwise the fiber app can not determine the real ip

@gaby  maybe we should extend the doc for these cases (ip method)

**Comment by ReneWerner87** [comment]
https://github.com/gofiber/fiber/blob/master/ctx_test.go#L1265

**Comment by gaby** [comment]
> https://github.com/gofiber/contrib/blob/bae3c8cc2db1c8cfcf747203af34f0ffa6a96bd9/otelfiber/semconv.go#L59-L62
> 
> https://github.com/gofiber/fiber/blob/634f163e3f6292e658e61d0dd9e3c475d87b5d54/ctx.go#L699-L701
> 
> https://docs.gofiber.io/next/api/fiber#config ![image](https://user-images.githubusercontent.com/7063188/224712866-2b6efe52-e7f7-461c-853a-73b234e3323c.png)
> 
> did you configure this header ? otherwise the fiber app can not determine the real ip
> 
> @gaby maybe we should extend the doc for these cases (ip method)

Agree, it's a bit confusing. From a otelfiber perspective using `c.IPs()` may be better since opentelemetry will auto-parse the list and only use the first IP which is the real client IP.

**Comment by ReneWerner87** [comment]
@bangbaew have you ever tested what you get when you configure the header of the proxy (mostly forwarded-for ) in your fiber app ?

**Comment by bangbaew** [comment]
> @bangbaew have you ever tested what you get when you configure the header of the proxy (mostly forwarded-for ) in your fiber app ?

If you mean have I tried logging from `C.IPs()` and `c.GetReqHeaders()`, I've tried them and the real IPs are shown in the fmt.Println, they both echo the X-Forwarded-For
If I send a request over Kong Gateway endpoint, it will log this
```
"X-Forwarded-For": "{my real public ip}, 10.8.26.4",
"X-Real-Ip": "10.8.26.4"
```
The local IP is Kong instance's ip

If I send a request directly, it will log this
```
"X-Forwarded-For": "{my real public ip}",
"X-Real-Ip": "{my real public ip}"
```

but both of them will log this in Jaeger UI
</html>
</body>

<body>
<html>

http.client_ip | 10.8.51.49
-- | --





You can see that the http.client_ip in Jaeger UI is the instance's local ip, not any of the forwarded IPs.

**Comment by ReneWerner87** [comment]
@bangbaew like this
```go
	app := fiber.New(fiber.Config{
		ProxyHeader: fiber.HeaderXForwardedFor,
	})
```

**Comment by bangbaew** [comment]
> > But I don't know how to configure the header of the proxy in my fiber app.
> 
> @bangbaew like this
> 
> ```go
> app := fiber.New(fiber.Config{
> 	ProxyHeader: fiber.HeaderXForwardedFor,
> })
> ```
> 
> https://docs.gofiber.io/next/api/fiber#config ![image](https://user-images.githubusercontent.com/7063188/224937124-2c999570-5b5d-4a86-a4c1-176d86e83d96.png)

Thanks a lot! it shows the X-Forwarded-For IPs now, with both public IP and API Gateway's IP, can I make it record only the first value?

**Comment by ReneWerner87** [comment]
do not think so, I would have to research

in any case, we should expand the documentation

@bangbaew  you can do that, you know best where you searched for the solution of the problem

**Comment by ReneWerner87** [comment]
https://github.com/gofiber/fiber/commit/0dee42a57cd76d7922a753d437894fa214819a63

**Comment by gaby** [comment]
@bangbaew opentelemetry says they only take the first value. Has that been the case for you after adding the header?

**Comment by ReneWerner87** [comment]
maybe we can change the middleware and cut away the second value which comes back through the header

**Comment by gaby** [comment]
@bangbaew What was the outcome of this?
